ibis/backends/tests/test_register.py

```diff
@@ -398,8 +399,8 @@ def test_register_garbage(con, monkeypatch):
 @pytest.mark.notyet(["impala", "mssql", "mysql", "postgres", "sqlite", "trino"])
 @pytest.mark.notimpl(
     ["flink"],
-    raises=ValueError,
-    reason="read_parquet() missing required argument: 'schema'",
+    raises=Py4JJavaError,
+    reason="Parquet format jar for Flink is not in the path",
```

Comment:
Is there a way we can get the parquet setup to work properly in CI? Right now this test doesn't check that flink is properly reading in the parquet file.

Reply:
We could `wget` the required jar file(s) in `ibis/docker/flink/Dockerfile`, and then call `add_jars()` in `TestConf` for Flink backend.

I tried to test this locally with
1. Download `flink-sql-parquet-1.17.2.jar` and `flink-parquet-1.18.1.jar` -- just to make sure we have the right jar file.
2. Add
```python
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-sql-parquet-1.17.2.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-parquet-1.18.1.jar'")
```
but got the following error
```bash
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o11.sqlQuery.
E                   : java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
E                       at org.apache.flink.formats.parquet.ParquetFileFormatFactory.getParquetConfiguration(ParquetFileFormatFactory.java:98)
E                       at org.apache.flink.formats.parquet.ParquetFileFormatFactory.access$000(ParquetFileFormatFactory.java:58)
E                       at org.apache.flink.formats.parquet.ParquetFileFormatFactory$ParquetBulkDecodingFormat.createRuntimeDecoder(ParquetFileFormatFactory.java:142)
E                       at org.apache.flink.formats.parquet.ParquetFileFormatFactory$ParquetBulkDecodingFormat.createRuntimeDecoder(ParquetFileFormatFactory.java:123)
E                       at org.apache.flink.connector.file.table.FileSystemTableSource.getScanRuntimeProvider(FileSystemTableSource.java:171)
E                       at org.apache.flink.table.planner.connectors.DynamicSourceUtils.validateScanSource(DynamicSourceUtils.java:478)
E                       at org.apache.flink.table.planner.connectors.DynamicSourceUtils.prepareDynamicSource(DynamicSourceUtils.java:161)
E                       at org.apache.flink.table.planner.connectors.DynamicSourceUtils.convertSourceToRel(DynamicSourceUtils.java:125)
E                       at org.apache.flink.table.planner.plan.schema.CatalogSourceTable.toRel(CatalogSourceTable.java:118)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.toRel(SqlToRelConverter.java:3997)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertIdentifier(SqlToRelConverter.java:2867)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertFrom(SqlToRelConverter.java:2427)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertFrom(SqlToRelConverter.java:2341)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertFrom(SqlToRelConverter.java:2286)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertSelectImpl(SqlToRelConverter.java:723)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertSelect(SqlToRelConverter.java:709)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3843)
E                       at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:617)
E                       at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.org$apache$flink$table$planner$calcite$FlinkPlannerImpl$$rel(FlinkPlannerImpl.scala:229)
E                       at org.apache.flink.table.planner.calcite.FlinkPlannerImpl.rel(FlinkPlannerImpl.scala:205)
E                       at org.apache.flink.table.planner.operations.SqlNodeConvertContext.toRelRoot(SqlNodeConvertContext.java:69)
E                       at org.apache.flink.table.planner.operations.converters.SqlQueryConverter.convertSqlNode(SqlQueryConverter.java:48)
E                       at org.apache.flink.table.planner.operations.converters.SqlNodeConverters.convertSqlNode(SqlNodeConverters.java:73)
E                       at org.apache.flink.table.planner.operations.SqlNodeToOperationConversion.convertValidatedSqlNode(SqlNodeToOperationConversion.java:272)
E                       at org.apache.flink.table.planner.operations.SqlNodeToOperationConversion.convert(SqlNodeToOperationConversion.java:262)
E                       at org.apache.flink.table.planner.delegation.ParserImpl.parse(ParserImpl.java:106)
E                       at org.apache.flink.table.api.internal.TableEnvironmentImpl.sqlQuery(TableEnvironmentImpl.java:708)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E                       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E                       at java.lang.reflect.Method.invoke(Method.java:498)
E                       at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E                       at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
E                       at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
E                       at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E                       at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
E                       at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
E                       at java.lang.Thread.run(Thread.java:748)
```

I will look more into this, and will also check with @chloeh13q and @deepyaman if they had any experience with this.

Reply:
It's been a while since I've done hadoop stuff, but it looks to me like the `hadoop-core` jar isn't installed or isn't on the classpath. This is likely something we'll need to change in the docker container.

If _needed_ I'd be fine with merging this without a functioning test _for now_, provided someone had at least verified somewhere that these `read_*` calls do work for the ibis flink backend. Having a functioning test setup would be better though.

Reply:
I have not run into this before. Have you checked if this classpath exists? From a quick Google search: https://stackoverflow.com/questions/29092926/java-lang-classnotfoundexception-org-apache-hadoop-conf-configuration

Reply:
I was able to clear the error above and a few others with the following list of JAR's:
```python
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-sql-parquet-1.18.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-parquet-1.18.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-hadoop-fs-1.18.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-hadoop-compatibility_2.12-1.18.1.jar'")

    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/hadoop-core-1.2.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/hadoop-common-3.3.6.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/hadoop-hdfs-3.3.6.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/hadoop-client-3.3.6.jar'")

    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/commons-logging-1.3.0.jar'")
```

but now hitting a strange issue with connecting to the Flink job manager:
```
E       pyflink.util.exceptions.TableException: org.apache.flink.table.api.TableException: Failed to execute sql
E               at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeQueryOperation(TableEnvironmentImpl.java:1084)
E               at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1120)
E               at org.apache.flink.table.api.internal.TableImpl.execute(TableImpl.java:477)
E               at org.apache.flink.table.runtime.arrow.ArrowUtils.collectAsPandasDataFrame(ArrowUtils.java:558)
E               at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E               at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E               at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E               at java.lang.reflect.Method.invoke(Method.java:498)
E               at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E               at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
E               at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
E               at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E               at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
E               at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
E               at java.lang.Thread.run(Thread.java:748)
E       Caused by: org.apache.flink.util.FlinkException: Failed to execute job 'collect'.
E               at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.executeAsync(StreamExecutionEnvironment.java:2253)
E               at org.apache.flink.table.planner.delegation.DefaultExecutor.executeAsync(DefaultExecutor.java:110)
E               at org.apache.flink.table.executor.python.ChainingOptimizingExecutor.executeAsync(ChainingOptimizingExecutor.java:88)
E               at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeQueryOperation(TableEnvironmentImpl.java:1065)
E               ... 14 more
E       Caused by: java.lang.RuntimeException: Error while waiting for job to be initialized
E               at org.apache.flink.client.ClientUtils.waitUntilJobInitializationFinished(ClientUtils.java:151)
E               at org.apache.flink.client.deployment.executors.AbstractSessionClusterExecutor.lambda$execute$2(AbstractSessionClusterExecutor.java:87)
E               at org.apache.flink.util.function.FunctionUtils.lambda$uncheckedFunction$2(FunctionUtils.java:73)
E               at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
E               at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591)
E               at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:457)
E               at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
E               at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
E               at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
E               at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
E       Caused by: java.util.concurrent.ExecutionException: org.apache.flink.util.concurrent.FutureUtils$RetryException: Could not complete the operation. Number of retries has been exhausted.
E               at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
E               at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
E               at org.apache.flink.client.deployment.executors.AbstractSessionClusterExecutor.lambda$null$0(AbstractSessionClusterExecutor.java:88)
E               at org.apache.flink.client.ClientUtils.waitUntilJobInitializationFinished(ClientUtils.java:131)
E               ... 9 more
E       Caused by: org.apache.flink.util.concurrent.FutureUtils$RetryException: Could not complete the operation. Number of retries has been exhausted.
E               at org.apache.flink.util.concurrent.FutureUtils.lambda$retryOperationWithDelay$6(FutureUtils.java:294)
E               at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
E               at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
E               at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
E               at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
E               at org.apache.flink.runtime.rest.RestClient.lambda$submitRequest$4(RestClient.java:590)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:583)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:559)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:629)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:118)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:321)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:337)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
E               at org.apache.flink.shaded.netty4.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
E               at java.lang.Thread.run(Thread.java:748)
E       Caused by: java.util.concurrent.CompletionException: org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8081
E               at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
E               at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
E               at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:957)
E               at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
E               ... 19 more
E       Caused by: org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8081
E       Caused by: java.net.ConnectException: Connection refused
E               at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
E               at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
E               at org.apache.flink.shaded.netty4.io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
E               at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
E               at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
E               at org.apache.flink.shaded.netty4.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
E               at java.lang.Thread.run(Thread.java:748)
```

The errors raised by the Flink container:
```
flink-1  | 2024-02-02 16:44:20,890 INFO  org.apache.flink.runtime.taskexecutor.DefaultJobLeaderService [] - Start job leader service.
flink-1  | 2024-02-02 16:44:20,891 INFO  org.apache.flink.runtime.filecache.FileCache                 [] - User file cache uses directory /tmp/flink-dist-cache-633f420c-c122-4289-94f1-730ae120d5d0
flink-1  | 2024-02-02 16:44:20,893 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Connecting to ResourceManager pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*(00000000000000000000000000000000).
flink-1  | 2024-02-02 16:44:21,023 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Resolved ResourceManager address, beginning registration
flink-1  | 2024-02-02 16:44:21,087 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Successful registration at resource manager pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_* under registration id 67b587afbd8c08e17da613fc7be789be.
flink-1  | 2024-02-02 16:44:53,482 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Disassociated]
flink-1  | 2024-02-02 16:45:30,243 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - The heartbeat of ResourceManager with id 695b26c25a0afb42b7ef44d6a9d32ce7 timed out.
flink-1  | 2024-02-02 16:45:30,249 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Close ResourceManager connection 695b26c25a0afb42b7ef44d6a9d32ce7.
flink-1  | 2024-02-02 16:45:30,255 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Connecting to ResourceManager pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*(00000000000000000000000000000000).
flink-1  | 2024-02-02 16:45:30,289 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:45:30,291 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:45:40,322 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:45:40,324 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:45:50,346 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:45:50,347 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:00,383 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:00,384 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:10,421 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:10,422 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:20,450 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:20,452 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:30,482 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:30,483 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:40,509 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:40,509 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:46:50,550 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:46:50,551 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:47:00,583 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:47:00,584 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:47:10,611 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:47:10,612 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
flink-1  | 2024-02-02 16:47:20,643 WARN  org.apache.pekko.remote.ReliableDeliverySupervisor           [] - Association with remote system [pekko.tcp://flink@flink-jobmanager:6123] has failed, address is now gated for [50] ms. Reason: [Association failed with [pekko.tcp://flink@flink-jobmanager:6123]] Caused by: [java.net.UnknownHostException: flink-jobmanager: Name or service not known]
flink-1  | 2024-02-02 16:47:20,644 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address pekko.tcp://flink@flink-jobmanager:6123/user/rpc/resourcemanager_*.
```

I start the Flink service(s) with
```bash
docker compose up flink
```
as usual.

Reply:
I confirmed that the container `ibis-flink-jobmanager` fails. Will try to pull its failure logs.

Reply:
By adding the JAR files in `flink/Dockerfile` as
```
FROM flink:1.18.1
# ibis-flink requires PyFlink dependency
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-python/1.18.1/flink-python-1.18.1.jar

RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-parquet/1.18.1/flink-parquet-1.18.1.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-sql-parquet/1.18.1/flink-sql-parquet-1.18.1.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-hadoop-fs/1.18.1/flink-hadoop-fs-1.18.1.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-hadoop-compatibility_2.12/1.18.1/flink-hadoop-compatibility_2.12-1.18.1.jar

RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-core/1.2.1/hadoop-core-1.2.1.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/3.3.6/hadoop-common-3.3.6.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-hdfs/3.3.6/hadoop-hdfs-3.3.6.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-client/3.3.6/hadoop-client-3.3.6.jar

RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/commons-logging/commons-logging/1.3.0/commons-logging-1.3.0.jar

RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-auth/3.3.6/hadoop-auth-3.3.6.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-hdfs-client/3.3.6/hadoop-hdfs-client-3.3.6.jar

RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/com/fasterxml/woodstox/woodstox-core/6.6.0/woodstox-core-6.6.0.jar
RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/codehaus/woodstox/stax2-api/4.2.2/stax2-api-4.2.2.jar
```

I was able to "narrow down" the error to:
```bash
$ docker compose up flink --force-recreate

flink-1  | 2024-02-02 18:17:39,754 WARN  org.apache.flink.runtime.util.HadoopUtils                    [] - Could not find Hadoop configuration via any of the supported methods (Flink configuration, environment variables).
flink-1  | 2024-02-02 18:17:39,755 ERROR org.apache.flink.runtime.taskexecutor.TaskManagerRunner      [] - Terminating TaskManagerRunner with exit code 1.
flink-1  | java.lang.NoClassDefFoundError: Could not initialize class org.apache.hadoop.security.UserGroupInformation
flink-1  | 	at org.apache.flink.runtime.security.modules.HadoopModule.install(HadoopModule.java:68) ~[flink-dist-1.18.1.jar:1.18.1]
flink-1  | 	at org.apache.flink.runtime.security.SecurityUtils.installModules(SecurityUtils.java:76) ~[flink-dist-1.18.1.jar:1.18.1]
flink-1  | 	at org.apache.flink.runtime.security.SecurityUtils.install(SecurityUtils.java:57) ~[flink-dist-1.18.1.jar:1.18.1]
flink-1  | 	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.runTaskManagerProcessSecurely(TaskManagerRunner.java:531) [flink-dist-1.18.1.jar:1.18.1]
flink-1  | 	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.runTaskManagerProcessSecurely(TaskManagerRunner.java:515) [flink-dist-1.18.1.jar:1.18.1]
flink-1  | 	at org.apache.flink.runtime.taskexecutor.TaskManagerRunner.main(TaskManagerRunner.java:473) [flink-dist-1.18.1.jar:1.18.1]
flink-1 exited with code 1
```

According to what I find online (e.g., [here](https://stackoverflow.com/questions/65157884/adding-hadoop-dependencies-to-standalone-flink-cluster)), we would need to set up Hadoop and point Flink environment to it in order to clear this error. My Hadoop knowledge is very limited, so I will put this on hold for now.

Reply:
**Update**: I found in an [outdated Flink doc](https://nightlies.apache.org/flink/flink-docs-release-1.10/ops/deployment/hadoop.html#adding-hadoop-to-lib) that Flink provides a bundled JAR for Hadoop integration as Hadoop requires many JAR's -- as validated/documented in my previous messages.

With the following list
```python
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-parquet-1.18.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-sql-parquet-1.18.1.jar'")
    con.raw_sql(f"ADD JAR '/Users/mehmet/Downloads/flink-shaded-hadoop-2-uber-2.8.3-10.0.jar'")
```

I was able to clear the previous errors. However, now getting
```
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o1770.hasNext.
E                   : java.lang.RuntimeException: Failed to fetch next result
E                       at org.apache.flink.streaming.api.operators.collect.CollectResultIterator.nextResultFromFetcher(CollectResultIterator.java:109)
E                       at org.apache.flink.streaming.api.operators.collect.CollectResultIterator.hasNext(CollectResultIterator.java:80)
E                       at org.apache.flink.table.planner.connectors.CollectDynamicSink$CloseableRowIteratorWrapper.hasNext(CollectDynamicSink.java:222)
E                       at org.apache.flink.table.runtime.arrow.ArrowUtils$1.hasNext(ArrowUtils.java:571)
E                       at org.apache.flink.table.runtime.arrow.ArrowUtils$2.hasNext(ArrowUtils.java:585)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E                       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E                       at java.lang.reflect.Method.invoke(Method.java:498)
E                       at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E                       at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
E                       at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
E                       at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E                       at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
E                       at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
E                       at java.lang.Thread.run(Thread.java:748)
E                   Caused by: java.io.IOException: Failed to fetch job execution result

...

E                   Caused by: org.apache.flink.util.FlinkException: Global failure triggered by OperatorCoordinator for 'Source: ibis_read_parquet_qdg5gsrw3jg67omm3gykbto2i4[1] -> Calc[$
E                       at org.apache.flink.runtime.operators.coordination.OperatorCoordinatorHolder$LazyInitializedCoordinatorContext.failJob(OperatorCoordinatorHolder.java:624)
E                       at org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$QuiesceableContext.failJob(RecreateOnResetOperatorCoordinator.java:248)
E                       at org.apache.flink.runtime.source.coordinator.SourceCoordinatorContext.failJob(SourceCoordinatorContext.java:395)
E                       at org.apache.flink.runtime.source.coordinator.SourceCoordinator.start(SourceCoordinator.java:225)
E                       at org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$DeferrableCoordinator.applyCall(RecreateOnResetOperatorCoordinator.java:326)
E                       at org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator.start(RecreateOnResetOperatorCoordinator.java:71)
E                       at org.apache.flink.runtime.operators.coordination.OperatorCoordinatorHolder.start(OperatorCoordinatorHolder.java:185)
E                       at org.apache.flink.runtime.scheduler.DefaultOperatorCoordinatorHandler.startOperatorCoordinators(DefaultOperatorCoordinatorHandler.java:165)
E                       at org.apache.flink.runtime.scheduler.DefaultOperatorCoordinatorHandler.registerAndStartNewCoordinators(DefaultOperatorCoordinatorHandler.java:159)
E                       at org.apache.flink.runtime.scheduler.adaptivebatch.AdaptiveBatchScheduler.initializeOperatorCoordinatorsFor(AdaptiveBatchScheduler.java:453)
E                       at org.apache.flink.runtime.scheduler.adaptivebatch.AdaptiveBatchScheduler.updateTopology(AdaptiveBatchScheduler.java:445)
E                       at org.apache.flink.runtime.scheduler.adaptivebatch.AdaptiveBatchScheduler.initializeVerticesIfPossible(AdaptiveBatchScheduler.java:311)
E                       at org.apache.flink.runtime.scheduler.adaptivebatch.AdaptiveBatchScheduler.startSchedulingInternal(AdaptiveBatchScheduler.java:179)
E                       at org.apache.flink.runtime.scheduler.SchedulerBase.startScheduling(SchedulerBase.java:625)
E                       at org.apache.flink.runtime.jobmaster.JobMaster.startScheduling(JobMaster.java:1075)
E                       at org.apache.flink.runtime.jobmaster.JobMaster.startJobExecution(JobMaster.java:992)
E                       at org.apache.flink.runtime.jobmaster.JobMaster.onStart(JobMaster.java:433)
E                       at org.apache.flink.runtime.rpc.RpcEndpoint.internalCallOnStart(RpcEndpoint.java:198)
E                       at org.apache.flink.runtime.rpc.pekko.PekkoRpcActor$StoppedState.lambda$start$0(PekkoRpcActor.java:618)
E                       at org.apache.flink.runtime.concurrent.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68)
E                       at org.apache.flink.runtime.rpc.pekko.PekkoRpcActor$StoppedState.start(PekkoRpcActor.java:617)
E                       at org.apache.flink.runtime.rpc.pekko.PekkoRpcActor.handleControlMessage(PekkoRpcActor.java:190)
E                       at org.apache.pekko.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:33)
E                       at org.apache.pekko.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:29)
E                       at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
E                       at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
E                       at org.apache.pekko.japi.pf.UnitCaseStatement.applyOrElse(CaseStatements.scala:29)
E                       at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
E                       ... 14 more
E                   Caused by: org.apache.flink.util.FlinkRuntimeException: Could not enumerate file splits
E                       at org.apache.flink.connector.file.src.AbstractFileSource.createEnumerator(AbstractFileSource.java:143)
E                       at org.apache.flink.runtime.source.coordinator.SourceCoordinator.start(SourceCoordinator.java:221)
E                       ... 38 more
E                   Caused by: java.io.FileNotFoundException: File /Users/mehmet/Desktop/ibis/ci/ibis-testing-data/parquet/functional_alltypes.parquet does not exist or the user running $
E                       at org.apache.flink.core.fs.local.LocalFileSystem.getFileStatus(LocalFileSystem.java:106)
E                       at org.apache.flink.connector.file.src.enumerate.NonSplittingRecursiveEnumerator.enumerateSplits(NonSplittingRecursiveEnumerator.java:83)
E                       at org.apache.flink.connector.file.src.AbstractFileSource.createEnumerator(AbstractFileSource.java:141)
E                       ... 39 more
```

In my understanding, this error is due to Flink task manager not being able to access the file path. This makes sense because `ibis/ci/ibis-testing-data/parquet/` is local and not accessible by Flink containers. However, we cannot also simply point to a file that is in flink-task-manager's file system either because we need to extract a schema from the file in `read_parquet()`. According to my current knowledge, the only viable solution seems to be having a file system shared between the Flink client (laptop or CI instance) and Flink containers.

Reply:
Verified that the issue is due to the fact that Flink client and containers are operate on different file systems:
- Switched to `local env` in `TestConf` for Flink backend:
```python
...

        stream_env = gateway.jvm.org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
        # flink_cluster_addr = os.environ.get("FLINK_REMOTE_CLUSTER_ADDR", "localhost")
        # flink_cluster_port = int(os.environ.get("FLINK_REMOTE_CLUSTER_PORT", "8081"))
        # j_stream_execution_environment = stream_env.createRemoteEnvironment(
        #     flink_cluster_addr,
        #     flink_cluster_port,
        #     env_settings.getConfiguration(),
        #     string_array,
        # )
        j_stream_execution_environment = stream_env.createLocalEnvironment(
            1, env_settings.getConfiguration(),
        )

        env = StreamExecutionEnvironment(j_stream_execution_environment)

...
```
- Run the test
```
$ pytest -m flink --showlocals --capture=no --color=yes 'ibis/backends/tests/test_register.py::test_read_parquet'

2024-02-05 17:48:12,645 INFO  org.apache.hadoop.io.compress.CodecPool                      [] - Got brand-new decompressor [.snappy]
..

============================= 2 passed, 38 deselected in 26.00s =============================
```

Reply:
> ... the only viable solution seems to be having a file system shared between the Flink client (laptop or CI instance) and Flink containers.

@jcrist Have you encountered a similar issue before? I could not think of a work-around for this. I keep coming back to the shared file system option. Though, I am not sure if that would be worth it to clear only one test function.